inst/include/dust/dust.hpp

```diff
@@ -521,6 +640,101 @@ class Dust {
     for (size_t i = 0; i < particles_.size(); ++i) {
       res[i] = particles_[i].compare_data(data[i / np], rng_.state(i));
     }
+    stale_device_ = true; // RNG use
+  }
+
+  std::vector<real_t> compare_data_device() {
+    std::vector<real_t> res;
+    auto d = device_data_offsets_.find(step());
+    if (d != device_data_offsets_.end()) {
+      res.resize(particles_.size());
+      compare_data_device(device_state_.compare_res, d->second);
+      device_state_.compare_res.get_array(res);
+    }
+    return res;
+  }
+
+  /*
+    This is exlcuded from test coverage, because initialise_device_data() will
+    be a noop in this case. This makes the find() above miss, and return NULL
+    to R without calling this
+  */
+  template <typename U = T>
+  typename std::enable_if<!dust::has_gpu_support<U>::value, void>::type
+  compare_data_device(dust::device_array<real_t>& res,                      // #nocov
+                      const size_t data_offset) {
+    throw std::invalid_argument("GPU support not enabled for this object"); // #nocov
+  }
+
+  template <typename U = T>
+  typename std::enable_if<dust::has_gpu_support<U>::value, void>::type
+  compare_data_device(dust::device_array<real_t>& res,
+                      const size_t data_offset) {
+    refresh_device();
+#ifdef __NVCC__
+    const int warp_size = dust::cuda::warp_size;
+    // Set up blocks and shared memory preferences
+    size_t blockSize = 128;
```

Comment:
this appears in enough places we should make it a `#define` in our cuda header?

Reply:
Although 128 is most common, this is generally different for each kernel, and the result of optimising with the profiler. I will give all of these different names however, to make this clearer